numpyro/contrib/control_flow/scan.py

```diff
@@ -187,24 +193,34 @@ def body_fn(wrapped_carry, x, prefix=None):
         site['infer']['dim_to_name'][time_dim] = '_time_{}'.format(first_var)
 
     # similar to carry, we need to reshape due to shape alternating in markov
-    ys = tree_multimap(lambda z0, z: jnp.reshape(z, z.shape[:1] + jnp.shape(z0)), y0, ys)
+    ys = tree_multimap(lambda z0, z: jnp.reshape(z, z.shape[:1] + jnp.shape(z0)[1:]), y0s, ys)
+    # then join with y0s
+    ys = tree_multimap(lambda z0, z: jnp.concatenate([z0, z], axis=0), y0s, ys)
     # we also need to reshape `carry` to match sequential behavior
-    if length % 2 == 0:
+    i = (length - 1) % (history + 1)
+    # XXX: unless we unroll more steps, we only know the correct shapes starting from
+    # the `history`-th iteration; that means we only know carry shapes when
+    # i == history - 1 or i == history, which corresponds to input carry or output carry
+    # at the `history`-th iteration.
```

Comment:
@fehiepsi I think I understand this issue, thanks for explaining.  How much compilation overhead does unrolling `2 * history` steps add in the HMM examples?  Unrolling `history` steps in exchange for incorrect output `carry` shapes is a very subtle optimization that only produces constant-time savings in program length, and may cause maintenance difficulties down the line, so it's important to be sure that it's worth it.

Reply:
Thinking about it again, I guess it won't take much compiling time because history is usually small (1 or 2). Let me address the issue in this PR.